File: iosApp/iosApp/viewmodel/OtpViewModel.swift
import Foundation
import SwiftUI
import Shared

@MainActor
class OtpViewModel: ObservableObject {
    @Published var otp: String = ""
    @Published var isLoading: Bool = false
    @Published var isListening: Bool = false
    @Published var isProcessing: Bool = false
    @Published var error: String?
    
    let phoneNumber: String
    
    private let preferencesRepository: PreferencesRepository
    private let authRepository: AuthRepository
    private let userRepository: UserRepository
    private let verifyOtpUseCase: VerifyOtpUseCase
    private let speechRepository: IOSSpeechRepository
    
    init(phoneNumber: String) {
        self.phoneNumber = phoneNumber
        self.preferencesRepository = IOSPreferencesRepository()
        self.authRepository = IOSAuthRepository(preferencesRepository: preferencesRepository)
        self.userRepository = FirestoreUserRepository(authRepository: authRepository)
        self.verifyOtpUseCase = VerifyOtpUseCase(authRepository: authRepository)
        self.speechRepository = IOSSpeechRepository()
    }
    
    func verifyOtp(
        onExistingUser: @escaping (String) -> Void, // userName callback for existing users
        onNewUser: @escaping () -> Void, // callback for new users
        onError: @escaping (String) -> Void
    ) async {
        guard otp.count == 6 else {
            error = "Please enter 6 digit OTP"
            return
        }
        
        isLoading = true
        error = nil
        
        do {
            try await verifyOtpUseCase.invoke(otp: otp)
            
            // Check if user exists in Firestore
            if let currentUser = try await userRepository.getCurrentUser() {
                // Existing user - show welcome back screen
                isLoading = false
                onExistingUser(currentUser.name)
            } else {
                // New user - continue to name screen
                isLoading = false
                onNewUser()
            }
        } catch {
            isLoading = false
            let errorMessage = (error as? Error)?.localizedDescription ?? "Invalid OTP"
            self.error = errorMessage
            onError(errorMessage)
        }
    }
    
    func resendOtp(onResend: @escaping () -> Void) {
        onResend()
    }
    
    func startSpeechRecognition() async {
        guard !isListening else { return }
        
        isListening = true
        error = nil
        
        // Haptic feedback
        let impactFeedback = UIImpactFeedbackGenerator(style: .medium)
        impactFeedback.impactOccurred()
        
        do {
            // Request permission if needed
            if !speechRepository.hasPermission() {
                let granted = try await speechRepository.requestPermission()
                if !granted.boolValue {
                    isListening = false
                    error = "Speech recognition permission is required"
                    return
                }
            }
            
            // Start listening with continuous updates
            Task {
                do {
                    try await speechRepository.startListeningWithUpdates { [weak self] recognizedText in
                        guard let self = self else { return }
                        // Extract digits from recognized text and limit to 6 digits for OTP
                        let digits = String(recognizedText.filter { $0.isNumber }.prefix(6))
                        if !digits.isEmpty {
                            self.otp = digits
                        }
                    }
                } catch {
                    if !error.localizedDescription.contains("cancelled") {
                        self.error = (error as? Error)?.localizedDescription ?? "Speech recognition error"
                    }
                    self.isListening = false
                }
            }
        } catch {
            self.error = (error as? Error)?.localizedDescription ?? "Speech recognition error"
            isListening = false
        }
    }
    
    func stopSpeechRecognition() async {
        guard isListening else { return }
        
        // Haptic feedback
        let impactFeedback = UIImpactFeedbackGenerator(style: .light)
        impactFeedback.impactOccurred()
        
        // Set processing state
        isListening = false
        isProcessing = true
        
        let initialText = speechRepository.stopListeningSync()
        // Extract digits and limit to 6 digits for OTP
        let digits = String(initialText.filter { $0.isNumber }.prefix(6))
        if !digits.isEmpty {
            otp = digits
        }
        
        // Wait a bit for final results to come in (up to 3 seconds)
        try? await Task.sleep(nanoseconds: 3_500_000_000) // 3.5 seconds
        
        // Get final accumulated text after waiting
        let finalText = speechRepository.getAccumulatedText()
        let finalDigits = String(finalText.filter { $0.isNumber }.prefix(6))
        if !finalDigits.isEmpty {
            otp = finalDigits
        }
        
        isProcessing = false
    }
}

File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/OtpViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.AndroidSpeechRepository
import com.kissangram.usecase.VerifyOtpUseCase
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class OtpViewModel(
    application: Application,
    val phoneNumber: String
) : AndroidViewModel(application) {
    
    private val preferencesRepository = AndroidPreferencesRepository(application)
    private val authRepository = AndroidAuthRepository(
        context = application,
        preferencesRepository = preferencesRepository
    )
    private val verifyOtpUseCase = VerifyOtpUseCase(authRepository)
    private val speechRepository = AndroidSpeechRepository(application)
    
    private val _uiState = MutableStateFlow(OtpUiState())
    val uiState: StateFlow<OtpUiState> = _uiState.asStateFlow()
    
    fun updateOtp(otp: String) {
        _uiState.value = _uiState.value.copy(
            otp = otp,
            error = null
        )
    }
    
    fun verifyOtp(onSuccess: () -> Unit, onError: (String) -> Unit) {
        val otp = _uiState.value.otp
        if (otp.length != 6) {
            _uiState.value = _uiState.value.copy(
                error = "Please enter 6 digit OTP"
            )
            return
        }
        
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            
            try {
                verifyOtpUseCase(otp)
                _uiState.value = _uiState.value.copy(isLoading = false)
                onSuccess()
            } catch (exception: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = exception.message ?: "Invalid OTP"
                )
                onError(exception.message ?: "Invalid OTP")
            }
        }
    }
    
    fun resendOtp(onResend: () -> Unit) {
        onResend()
    }
    
    fun startSpeechRecognition() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isListening = true, error = null)
            
            // Set up callback for continuous updates BEFORE starting to listen
            speechRepository.setOnTextUpdate { recognizedText ->
                // Extract digits from recognized text and update continuously
                // Limit to 6 digits for OTP
                val digits = recognizedText.filter { it.isDigit() }.take(6)
                if (digits.isNotEmpty()) {
                    _uiState.value = _uiState.value.copy(
                        otp = digits,
                        error = null
                    )
                }
            }
            
            // Start listening in background - it will continue until stopped
            try {
                speechRepository.startListening()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isListening = false,
                    error = e.message ?: "Speech recognition error"
                )
            }
        }
    }
    
    fun stopSpeechRecognition() {
        android.util.Log.d("OtpViewModel", "stopSpeechRecognition called")
        viewModelScope.launch {
            // Set processing state
            _uiState.value = _uiState.value.copy(isListening = false, isProcessing = true)
            android.util.Log.d("OtpViewModel", "Calling speechRepository.stopListening()")
            speechRepository.stopListening()
            // Wait a bit for final results to come in (up to 3 seconds)
            delay(3500) // Wait 3.5 seconds to ensure final results are processed
            // Get final accumulated text
            val finalText = speechRepository.getAccumulatedText()
            val digits = finalText.filter { it.isDigit() }.take(6)
            if (digits.isNotEmpty()) {
                _uiState.value = _uiState.value.copy(otp = digits)
            }
            _uiState.value = _uiState.value.copy(isProcessing = false)
        }
    }
}

data class OtpUiState(
    val otp: String = "",
    val isLoading: Boolean = false,
    val isListening: Boolean = false,
    val isProcessing: Boolean = false,
    val error: String? = null
)
